File: js/leetcode/_15.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
export default function sol(nums) {
  function helper(start, end) {
    if (start > end) return null;

    let mid = (start + (end - start) / 2) | 0;
    let root = {};
    root.val = nums[mid];
    root.left = helper(start, mid - 1);
    root.right = helper(mid + 1, end);
    return root;
  }
  return helper(0, nums.length - 1);
}

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

File: src/leetcode/_15.rs
#![allow(dead_code)]
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::rc::Rc;
struct Solution;
impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        fn helper(nums: &Vec<i32>, start: usize, end: usize) -> Option<Rc<RefCell<TreeNode>>> {
            let mid = start + (end - start) / 2;

            if start > end || mid >= nums.len() {
                return None;
            }

            let root = Rc::new(RefCell::new(TreeNode::new(nums[mid])));
            root.borrow_mut().left = helper(nums, start, mid - 1);
            root.borrow_mut().right = helper(nums, mid + 1, end);

            return Some(root);
        }
        return helper(&nums, 0, nums.len() - 1);
    }
}

#[cfg(test)]
mod test {
    use super::*;
}
